component/http/handler.go

```diff
@@ -62,44 +62,77 @@ func determineEncoding(r *http.Request) (string, encoding.DecodeFunc, encoding.E
 	var enc encoding.EncodeFunc
 	var dec encoding.DecodeFunc
 	var ct string
+	var err error
 
 	if cok {
-		switch cth[0] {
-		case "*/*", json.Type, json.TypeCharset:
-			enc = json.Encode
-			dec = json.Decode
-			ct = json.TypeCharset
-		case protobuf.Type, protobuf.TypeGoogle:
-			enc = protobuf.Encode
-			dec = protobuf.Decode
-			ct = protobuf.Type
-		default:
-			return "", nil, nil, errors.New("content type header not supported")
+		ah := getMultiValueHeaders(cth[0])
+		for _, v := range ah {
+			ct, dec, enc, err = getSingleHeaderEncoding(v)
+			if err == nil {
+				break
+			}
+		}
+		// No valid headers found after going through all headers
+		if err != nil {
+			return "", nil, nil, errors.New("accept header not supported")
 		}
+
 	}
 
 	if aok {
-		switch ach[0] {
-		case "*/*", json.Type, json.TypeCharset:
-			enc = json.Encode
-			if dec == nil {
-				dec = json.Decode
-			}
-			ct = json.TypeCharset
-		case protobuf.Type, protobuf.TypeGoogle:
-			enc = protobuf.Encode
-			if dec == nil {
-				dec = protobuf.Decode
+		ah := getMultiValueHeaders(ach[0])
+		for _, v := range ah {
+			ct, dec, enc, err = getSingleHeaderEncoding(v)
+			if err == nil {
+				break
 			}
-			ct = protobuf.Type
-		default:
+		}
+		// No valid headers found after going through all headers
+		if err != nil {
 			return "", nil, nil, errors.New("accept header not supported")
 		}
 	}
 
 	return ct, dec, enc, nil
 }
 
+func getSingleHeaderEncoding(header string) (string, encoding.DecodeFunc, encoding.EncodeFunc, error) {
+
+	var enc encoding.EncodeFunc
+	var dec encoding.DecodeFunc
+	var ct string
+
+	switch header {
+	case "*/*", json.Type, json.TypeCharset:
+		enc = json.Encode
+		dec = json.Decode
+		ct = json.TypeCharset
+	case protobuf.Type, protobuf.TypeGoogle:
+		enc = protobuf.Encode
+		dec = protobuf.Decode
+		ct = protobuf.Type
+	default:
+		return "", nil, nil, errors.New("accept header not supported")
+	}
+
+	return ct, dec, enc, nil
+}
+
+func getMultiValueHeaders(header string) []string {
+	if !strings.Contains(header, ",") {
+		return []string{header}
```

Comment:
We could not check for Contains and do split directly. Would that work?

Reply:
oh you mean skip `Contains` and call `Split` directly? something like:
```
func getMultiValueHeaders(header string) []string {
  splitHeaders := strings.Split(header, ",")
  if splitHeaders[0] == header {
    return []string{header}
  }
  ....
```

Reply:
yes, the cost of doing split vs contains might the same...maybe a benchmark could shed light if we want to be data-driven...
benchmark with single and multi heder with and without contains...

Reply:
i did some benchmark on that ...
```
GOROOT=/usr/local/go #gosetup
GOPATH=/Users/vangeliskatikaridis/go #gosetup
/usr/local/go/bin/go test -c -o /private/var/folders/f1/w8ll0_cj1bl_prh_f928f0zr0000gp/T/___gobench_github_com_beatlabs_patron_component_http github.com/beatlabs/patron/component/http #gosetup
/private/var/folders/f1/w8ll0_cj1bl_prh_f928f0zr0000gp/T/___gobench_github_com_beatlabs_patron_component_http -test.v -test.bench . -test.run ^$ #gosetup
goos: darwin
goarch: amd64
pkg: github.com/beatlabs/patron/component/http
BenchmarkSingleHeaderContains
BenchmarkSingleHeaderContains-16    	34828395	        30.6 ns/op
BenchmarkMultiHeaderContains
BenchmarkMultiHeaderContains-16     	 6916483	       173 ns/op
BenchmarkSingleHeaderSplit
BenchmarkSingleHeaderSplit-16       	16266246	        72.5 ns/op
BenchmarkMultiHeaderSplit
BenchmarkMultiHeaderSplit-16        	 7079029	       167 ns/op
PASS
```

Seems in the case of single value headers, the performance is better (by a factor of around 2) . 
For the multi value case , as probably expected, the performance is very much comparable.
In absolute values, the gain of using Contains for single headers is around 40ns

@astrikos , @mantzas  . What do you think ?

attaching my test class as reference

```
package http

import (
	"gotest.tools/assert"
	"strings"
	"testing"
)

func getMultiValueHeadersContains(header string) []string {
	if !strings.Contains(header, ",") {
		return []string{header}
	}

	splitHeaders := strings.Split(header, ",")

	trimmedHeaders := []string{}
	for _, v := range splitHeaders {
		trimmedHeaders = append(trimmedHeaders, strings.TrimSpace(v))
	}

	return trimmedHeaders
}

var headers []string

func getMultiValueHeadersSplit(header string) []string {
	splitHeaders := strings.Split(header, ",")

	trimmedHeaders := []string{}
	for _, v := range splitHeaders {
		trimmedHeaders = append(trimmedHeaders, strings.TrimSpace(v))
	}

	return trimmedHeaders
}

func TestSingleValueHeadersContains(t *testing.T) {

	headers := getMultiValueHeadersContains("string")

	assert.Equal(t,[]string{"string"},headers)

}

func TestMultiValueHeadersContains(t *testing.T) {

	headers := getMultiValueHeadersContains("string,string")

	assert.Equal(t,[]string{"string"},headers)

}

func TestSingleValueHeadersSplit(t *testing.T) {

	headers := getMultiValueHeadersContains("string")

	assert.Equal(t,[]string{"string","string"},headers)

}

func TestMultiValueHeadersSplit(t *testing.T) {

	headers := getMultiValueHeadersContains("string,string")

	assert.Equal(t,[]string{"string","string"},headers)

}

func BenchmarkSingleHeaderContains(b *testing.B){

	benchmarkMultiValueHeadersContains("string",b)

}

func BenchmarkMultiHeaderContains(b *testing.B){

	benchmarkMultiValueHeadersContains("string,string",b)

}


func BenchmarkSingleHeaderSplit(b *testing.B){

	benchmarkMultiValueHeadersSplit("string",b)

}

func BenchmarkMultiHeaderSplit(b *testing.B){

	benchmarkMultiValueHeadersSplit("string,string",b)

}

func benchmarkMultiValueHeadersContains(header string, b *testing.B) {
	for n := 0; n < b.N; n++ {
		headers = getMultiValueHeadersContains(header)
	}
}

func benchmarkMultiValueHeadersSplit(header string, b *testing.B) {
	for n := 0; n < b.N; n++ {
		headers = getMultiValueHeadersSplit(header)
	}
}


```

Reply:
I am happy with `Contains` for this case to be honest, it's comparable and I think a bit more readable. 
@mantzas if you are okay with that I am okay to move on and merge that.